### DISCUSSION TITLE
preserving comments

### DISCUSSION THREAD

**Discussion by vzaliva** [tree-sitter/tree-sitter#2868] [Q&A] [answered]
Hi! I am trying to parse a language where comments contain annotations which I want to preserve.  Now I have:

```
    extras: $ => [
        /\s/,
        /--.*\n/,
        /{-.*-}/,
    ],
```

How can I preserve my `{- ... -}` comments in the AST? They could appear anywhere where whitespace is allowed. Thanks!

**Comment by WillLillis** [comment, marked as answer]
Would it be possible to define the annotations as a subfield of a comment? Looking at what you currently have in the repo, would something like this work?

```js
module.exports = grammar({
    name: 'core',

    extras: $ => [
        /\s/,
        $.comment,
    ],

    
    rules: {
        source_file: $ => repeat1($.declaration),

        comment: _ => token(choice(
            /--.*\n/,
            seq('{-', field('annotation', /.*/), '-}'))),

...
```

**Reply by vzaliva** [reply]
Thanks!  I think this would work.
